### ISSUE TITLE
Allow Fibers to have a priority

### ISSUE THREAD

**Issue by disposableme** [puniverse/quasar#195]
I'd like to request a feature.  Allowing fibers to set priorities would be, in my opinion, a wonderful enhancement.

This is a major feature that threads have over fibers.
From the [User Manual](http://docs.paralleluniverse.co/quasar/#fibers):
> Fibers provide functionality similar to threads, and a similar API, but they’re not managed by the OS.

Setting a priority is a major feature of threads over fibers.  If this is added to the fiber (and strand) API, this would make fibers even more powerful.

**Edit:** To clarify, since Fibers use a non-preemptive scheduler and are OS independent, I am suggesting to only use the same API as threads, but not the same (OS dependent) prioritization approach as threads.  I think the most straight forward approach may be to implement priorities such that:

* Any fiber with a high priority will be started on a worker thread before (or at the same time as) a fiber with a low priority.
* Any already running fibers with a low priority will remain running until they block as normal.
* The priority should only dictate the order in which to run queued fibers, and will not affect already running fibers in any way.
* Priorities can be changed at any time and by anyone with a reference to the Fiber.
* Any change in priority will not have any affect until the next time a fiber blocks and a new fiber is selected to run.
* By default, a fiber will have the same priority as the strand that spawned it.

**Comment by disposableme** [comment]
I've been trying to think of a good workaround to the lack of this feature, but I can't think of a very elegant one.  The best that I've come up with is just using a separate priority queue to start fibers.  Just before blocking, each running fiber would need to check if Quasar's underlying ForkJoinPool has any queued fibers.  If there are no fibers queued to run, and therefore will result in an empty worker thread, then the priority queue should be popped and the next batch of high priority fibers should be started.

This workaround is not ideal for many reasons.  For example, the check for queued fibers needs to be done just before every blocking call, but this is not possible when the blocking call is done in a 3rd party library, such as in Comsat.

Unfortunately, since ForkJoinTasks don't have priorities, I assume that this feature probably won't be a trivial change to the fiber API.

**Comment by pron** [comment]
There are really two separate issues here: 

1. Allow fibers to have a priority
2. Respect those priorities in the various schedulers

1 is easy, of course, and is similar to thread priority. OSes are free to ignore thread priority, and I think (though I'm not at all sure) that some/many/most do, and those that do respect it, don't really specify how (or, at least, the JVM doesn't guarantee anything). Basically, thread priority is only required to be respected in real-time OSes, it has a clear semantics in those OSes, and it comes at a cost, which realtime applications gladly pay (realtime often sacrifices performance for predictability).

I guess it is possible to write a fiber scheduler that respects fiber priority in some way (although that needs to be specified), but I wonder what would be the use case. Normally, thread priority involves some tricky issues, like priority inversion (and how that's handled is well specified in realtime environments), and those would need to be addressed.

**Comment by disposableme** [comment]
Great points.

To handle 2, I assume that a new `FiberScheduler` would need to be created that was "priority aware".

I can't speak for other use cases, but my use case might be a bit complex.  I have many actors each running on their own fiber.  My program has multiple logical iterations.  For each iteration, every actor will run exactly once.  Each actor has a set of actors that it is dependent on and it is run after it receives a message from all of them.  (In other words, a selective `receive` that blocks until all of the required messages arrive.)

Let's say that I have 2 actors each running on fibers, H and L, with high and low priorities respectively.  While waiting to be run, they each receive a message that fulfills their dependencies such that each of their selective `receive` methods will return a value.  When a currently running fiber blocks, the scheduler must now decide which fiber to run next, with H and L both being candidates.  (I assume that the underlying default `ForkJoinPool` is just using a FIFO schedule to make this decision.)  I'd like H to be executed (or at least preferred to be executed) before L in this case.

There are tricky issues, as you mentioned, such as preventing the starvation of L.  In my use case, since I have iterative phases, I will want L from phase 0 to execute before H from phase 1.  Assuming I'm using a single core for simplicity, I would want to avoid an execution ordering that causes starvation such as `H0, H1, H2, H3, L0, L1, L2, L3` and instead have `H0, L0, H1, L1, H2, L2, H3, L3`.  This type of starvation could occur if high priority fibers were always run before low priority ones.

* Although not very elegant, this could be fixed by simply incrementing the priority of a fiber after each time it is run.
* A better approach to prevent this type of starvation may be to have a count marking the number of times an actor was run (after a successful call to `receive` or `tryReceive`) and giving preference to those with a low count while ties are broken with the priority.
* Another approach may be to simply always run the actor with the highest priority and just rely on the selective `receive` blocking to prevent starvation.  If the dependencies are interconnected enough between actors, the starvation issue mentioned above would probably not happen often.

I agree that there are many tricky issues regarding priority and it can be implemented in many different ways.  There's definitely not a one size fits all solution.

Also it may be worth mentioning that I keep saying "priority" but "niceness" may be more applicable.

**Comment by disposableme** [comment]
One big motivation for priorities would be to help the liveness in a distributed setting.

Assume Quasar running in a distributed setting with Galaxy.  Let's say there are 2 machines, A and B, on the network.  Each machine has 100 actors (A1-A100 and B1-B100).

If all actors on machine B have a selective `receive` that depends on A1, then we may run into liveness issues.  It may be prudent to give A1 a higher priority than the other actors on machine A.

On a single machine, this is probably not much of an issue, but in a distributed setting, this can potentially cause a large workload imbalance that will greatly hurt performance.  Since one machine has no information about the dependencies of the other, in order to schedule actors optimally, it probably needs to be left to the programmer via a mechanism such as fiber priority.

**Comment by linkerlin** [comment]
Use two or more ThreadPool Executors to schedule Fibers, then set diff-priorities to Executors.
Maybe a simple solution.

**Comment by disposableme** [comment]
@linkerlin I was originally thinking along those lines at first, but there are some issues.  That works fine if there are only a couple of different priorities, but unfortunately I don't think that it scales when there are many fibers with many different priorities.  If many ThreadPools are used, then the benefit of using light weight fibers is lost since the many threads will need to context switch.

**Comment by pron** [comment]
First let me say that I am not at all opposed to the idea of a prioritized scheduler. However, it is not simple to balance the expected utility with the inevitable performance hit. As I said, realtime schedulers have no choice; they must guarantee predictability at all cost. Soft-realtime application, however, are reluctant to lose performance.

@disposableme , your design for a priority scheduler would incur a very severe performance hit. Every worker must ensure that no higher priority job is waiting anywhere before running each and every task. In a work-stealing scheduler, this implies scanning all queues before every task selection; in a single-queue design, this implies contention on every insertion as well as every poll.

In that case, I think that a clever use of a `Phaser` would get you what you want, with a performance hit that isn't worse than your design. Each of your actors would `awaitAdvance` before scheduling (i.e. sending a message to) an actor in the next phase. However, an actor could `arrive` at the very beginning of the phase, because you don't mind next-phase executions as long as the current phase are already running (i.e., you don't need to wait for them to finish). There is a fiber-friendly Phaser in `co.pu.strands.concurrent`.

**Comment by voidstarstar** [comment]
Just to add my 2 cents...

A prioritized scheduler seems very useful.  Currently a `ForkJoinPool` only allows tasks to be executed in a FIFO ordering, so if there are many tasks in the queue, the user has little control over the order in which they're run.

Ideally, the scheduler would logically operate as follows:
1. An event occurs that causes fiber X to become unblocked.
2. Fiber X is submitted to the scheduler.  Anything submitted to the scheduler is able to run at any time.
3. When the scheduler decides which task to execute next, it chooses the one with the highest priority rather than the oldest one submitted.

*The priorities are simply loose, unenforced suggestions and no actual guarantees are made about the ordering.*  It should be allowable, however unlikely, that low priority tasks are executed before high priority ones.  For example, if we have a slow thread and a fast thread, the fast thread may finish 100 low priority tasks while the slow thread is still trying to complete 1 high priority task.  Priorities should NOT be depended upon to prevent deadlocks, livelocks, etc.  They should only be used for efficiency purposes (i.e. the program will terminate faster overall if high priority tasks tend to be run prior to lower priority tasks).


My opinion of the solutions offered so far:

1. The solution suggested by @disposableme to use a priority queue is an ugly hack.  The scheduling of fibers should not be exposed outside of the `FiberScheduler`.  The original suggestion of a new `FiberScheduler` is much better, but unclear on how to implement.
2. The solution suggested by @linkerlin is insufficient to solve the original problem for the reasons mentioned by @disposableme.
3. The solution suggested by @pron is unclear to me.  Can you give more detail on the use of a `Phaser`?  Would this still work for arbitrary priorities?  For example, would this still work if a `Comparator` were used to determine priorities?  This is currently a bit too unclear for me to know if it will work logically the same as a central priority queue.  Either way, it is still exposed beyond the `FiberScheduler`, which seems like a code smell to me.

I think the best option would be a modification of a `ForkJoinPool`.  From what I understand, there was a major change in the way a `ForkJoinPool` works from Java 7 to Java 8.

In Java 7:
> Submissions from non-FJ threads enter into **a submission queue**. Workers take these tasks and typically split them into subtasks that may be stolen by other workers. Preference rules give first priority to processing tasks from their own queues (LIFO or FIFO, depending on mode), then to randomized FIFO steals of tasks in other worker queues, **and lastly to new submissions**.

In Java 8:
> Submissions from non-FJ threads enter into **submission queues**. Workers take these tasks and typically split them into subtasks that may be stolen by other workers. Preference rules give first priority to processing tasks from their own queues (LIFO or FIFO, depending on mode), then to randomized FIFO steals of tasks in other queues.

(Emphasis mine.)

It seems like the central submission queue was removed and the tasks would be submitted directly to the worker thread queues for efficiency purposes.  Less contention on this queue means higher throughput of `ForkJoinTask`s, so this was a natural change.  However, this lack of a central queue means prioritization is more difficult.  Would it be possible to create a slightly modified `ForkJoinPool` based on the Java 7 version that uses a priority queue as the submission queue?  Perhaps the worker queues should be priority queues as well, but I'm not entirely sure.

The priority queue consumers would require a lock on the queue to remove tasks.  If the fibers are long running (with respect to the time the priority queue is locked), then I would expect very little contention on the priority queue.  On the other hand, if the fibers are short running, then I would expect a performance hit due to one worker thread locking the queue while another thread blocks while waiting on this lock.  I assume that multiple tasks can be popped at a time from the queue to also help alleviate the contention.  (It would be interesting to see if a typical fiber use case is even affected by this central queue contention or if the `ForkJoinPool` change from Java 7 to 8 had negligible impact when used in quasar.  If there is an impact on performance, then an additional disclaimer should be added, however it's not like the priority queue operations are `O(1)` anyway, so anyone who uses a prioritized `ForkJoinPool` should expect a performance hit as a trade-off.  It would be up to the user to determine if this trade-off is worth the benefit of prioritizing the fibers.)


Either way, I personally think the best solution would be a custom `FiberScheduler` instead of the previously proposed solutions since this does not expose scheduling to the user which would greatly increase the complexity.

**Comment by voidstarstar** [comment]
Do you guys think that something as simple as the following code would work?  The `Comparator` would just need to be created based on how you want to sort the fibers.  As I mentioned in my last post, I think this would cause the worker threads to block when contending for the lock on the `PriorityBlockingQueue`, but if the fibers run much longer than the lock time, then I don't think they would block very often.

```Java
int parallelism = Runtime.getRuntime().availableProcessors();
Comparator<? super Runnable> comparator = null;
PriorityBlockingQueue<Runnable> priorityBlockingQueue = new PriorityBlockingQueue<Runnable>(11, comparator);
ThreadPoolExecutor threadPool = new ThreadPoolExecutor(parallelism, parallelism, Long.MAX_VALUE,
				TimeUnit.NANOSECONDS, priorityBlockingQueue);
FiberScheduler fiberScheduler = new FiberExecutorScheduler("threadPool", threadPool);
```

Also, how can we force an actor to use this `FiberScheduler`?  I'm probably missing something obvious. :confused:

**Comment by circlespainter** [comment]
@voidstarstar You should be able to use [this overload](http://docs.paralleluniverse.co/quasar/javadoc/co/paralleluniverse/actors/Actor.html#spawn-co.paralleluniverse.fibers.FiberFactory-) (or [this one](http://docs.paralleluniverse.co/quasar/javadoc/co/paralleluniverse/actors/Actor.html#spawn-co.paralleluniverse.strands.StrandFactory-)) of `spawn` because `FiberExecutorScheduler` implements both `FiberFactory` and `StrandFactory`.

**Comment by voidstarstar** [comment]
Thanks @circlespainter!  I completely overlooked the `spawn` method - I was wrongly assuming that it would be set in the constructor.

Now it looks like the only issue is the `Comparator`.  It needs to be a `Comparator<? super Runnable>`, but we only have access to the `Actor` which does not implement `Runnable`.  Is the `Runnable` exposed via the API anywhere?

Is [this](https://github.com/puniverse/quasar/blob/master/quasar-core/src/main/java/co/paralleluniverse/strands/Strand.java#L790) where the `Runnable` is created that eventually gets passed to the `ForkJoinPool`?  If so, is there any reason why static nested wrapper classes (`SuspendableRunnableRunnable` and `SuspendableCallableRunnable`) are used instead of implementing the `Runnable` interface directly?

**Comment by pron** [comment]
> Would this still work for arbitrary priorities?

No, it is just a means to address @disposableme 's use case.

> I think the best option would be a modification of a ForkJoinPool

Absolutely. An approximate priority ordering, e.g. in each thread's queue is possible. However, I don't think that this is such a trivial change.

What we'll do is add `setPriority`/`getPriority` to the `Fiber` and `Strand` classes (the only cost is an additional 4 bytes to each fiber), but postpone writing a priority-based scheduler. This will allow anyone to write such a scheduler to suit their particular requirements. If anyone implements a sufficiently general and useful priority scheduler and wishes to contribute it, we'll be happy to merge it into Quasar core.

Irrespective of this, we will provide a fiber-compatible implementation of `PriorityBlockingQueue`.

**Comment by voidstarstar** [comment]
Thanks @pron!

I'm trying to write a scheduler using the new API changes, but I had a few questions.  If the new code is still a work in progress, please disregard this post.

1. The `Strand` uses an `int` but the `Fiber` uses a `byte`.  Was this intentional?  You had mentioned an additional 4 bytes per fiber, so was this later later determined to be too much overhead?
2. What is the purpose of setting `MIN_PRIORITY` and `MAX_PRIORITY` to numbers other than `Integer.MIN_VALUE` and `Integer.MAX_VALUE`?
3. What is the suggested way to access this new priority?  Are users supposed to create a `FiberExecutorScheduler` and pass a custom `Executor` to it?  If so, how does the `Executor` access the priority of a `Fiber`?  From what I understand, the `Executor` does not have access to the `Fiber` directly and the `RunnableFiberTask` class is not public, so type casting can not be done either.

(I had to modify the visibility of some classes/fields in quasar and change the min/max priority values in order to get a quick and dirty scheduler to work, but so far it looks like it works as expected!)

Thanks again for the great work!

**Comment by pron** [comment]
1. Because a byte is big enough to hold the range MIN_PRIORITY...MAX_PRIORITY
2. That's because those are the limits on `Thread` priority, and threads are strands and fibers are also strands.
3. You're right, there's a missing piece. The task passed to the scheduler is of type `co.pu.fibers.FiberTask`, which should be made public, and have a `getFiber` method. I'll do it right away.

**Comment by pron** [comment]
Alright, now you can access the fiber (and its priority) by casting the `Runnable` to `co.pu.fibers.FiberSchedulerTask`.

**Comment by voidstarstar** [comment]
1. I agree that a byte is large enough to hold the current range from the [1](https://github.com/puniverse/quasar/blob/9d0a7799e40c3cd687d1540970cb2a34f9ecd10a/quasar-core/src/main/java/co/paralleluniverse/strands/Strand.java#L61) to the [10](https://github.com/puniverse/quasar/blob/9d0a7799e40c3cd687d1540970cb2a34f9ecd10a/quasar-core/src/main/java/co/paralleluniverse/strands/Strand.java#L71), but isn't this an arbitrary limit that acts as an unnecessary restriction?  Shouldn't the range include all possible values that the priority field can be?  I would personally prefer using an `int` to allow for a much wider range in priorities, just as the OS scheduler does, but this of course has a slight increase in memory overhead.  A large range is useful since there are often many fibers.  A byte only allows for 256 possible values, so it would be quite restrictive if using millions of fibers.  Regardless of what primitive is used, I believe the choices should be one of the following options (and in all cases, the `NORM_PRIORITY` should be set to 0):
  * Use an `int`.  Set the range as `Integer.MIN_VALUE` to `Integer.MAX_VALUE`.
  * Use a `short`.  Set the range as `Short.MIN_VALUE` to `Short.MAX_VALUE`.
  * Use a `byte`.  Set the range as `Byte.MIN_VALUE` to `Byte.MAX_VALUE`.
2. See above.
3. Tested and confirmed to work great.  Thanks!

I'd be happy to submit a pull request to address the above issues if you'd like.

**Comment by pron** [comment]
> but isn't this an arbitrary limit that acts as an unnecessary restriction?

It is, but I didn't set this limit. Threads are strands and they already have this restriction which will cause `setPriority` to throw an exception if set outside the range.

> just as the OS scheduler does

No, Java threads have only 10 priorities.

**Comment by pron** [comment]
We could add instance methods to `Strand` to get the min and max priorities, and so allow fibers and threads to have different ranges, but before complicating things, let's see how useful this is.

**Comment by voidstarstar** [comment]
> It is, but I didn't set this limit. Threads are strands and they already have this restriction which will cause `setPriority` to throw an exception if set outside the range.

> No, Java threads have only 10 priorities.

Wow, I actually had no idea that Java was this restrictive!  I wonder why they even bother to use an `int` if there are only 10 possible values.  I'll have to look into this more, but I'm guessing that this is due to compatibility with a wide array of platforms supported by the JVM.

> We could add instance methods to Strand to get the min and max priorities, and so allow fibers and threads to have different ranges, but before complicating things, let's see how useful this is.

So both `Fiber`s and `Thread`s are `Strand`s, but only the `Fiber` is a subclass of `Strand`?  I ask because the priority restrictions seem like they really should belong to just a `Thread` rather than to all `Strand`s.

Instance methods would also work, or in the mean time, Fibers could simply ignore the limits (since the limits are only intended for `Thread`s anyway).

**Comment by pron** [comment]
Thing is, adding more priorities can make implementations of efficient schedulers much harder. If you have 10 priorities, you can implement a scheduler with 10 simple FIFO queues; if you have 2 billion -- you can't and must use some sorted queue, which increases contention heavily. 

In any event, let's start with this, and if there are compelling reasons to change this, we will.

**Comment by voidstarstar** [comment]
Ok, I see where you're coming from now.  I was originally trying to use a modified `ForkJoinPool`, but it was complicated to modify, so I decided to first try a simple `ThreadPoolExecutor`.

> you can't and must use some sorted queue, which increases contention heavily.

I'm using a sorted queue, but I think that the contention may end up being quite low in my use case.

I'm using a `ThreadPoolExecutor` with a `PriorityBlockingQueue` which [uses a lock](http://grepcode.com/file/repository.grepcode.com/java/root/jdk/openjdk/6-b14/java/util/concurrent/PriorityBlockingQueue.java?av=f#242) when a thread tries to `take` an element from it.  I think it should be relatively fast for ThreadA to acquire and release the lock (if the lock is free of course).  I think that thread-blocking will only occur if ThreadB calls the `take` method and attempts to acquire the lock while ThreadA is still within this critical section.  This contention is normally bad of course, but I think that there is a low probability of the lock blocking as long as the Fibers are long running tasks with respect to the lock time.  (Note: This does not mean that the Fibers are taking so much time that Threads should be used.)

My intuition is that short running Fibers (w.r.t. lock time) will cause high contention but long running Fibers (w.r.t. lock time) will cause low contention.  I may be completely wrong about this, so I will have to do some further testing to verify, but I believe that my use case has long running Fibers and so it will hopefully result in low contention.

As you said, in the mean time we can start with this.  (It's not a huge deal since I can always fall back to a custom fork of quasar if there's not a compelling reason to accept the changes upstream.)
